src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,315 @@
+
+    if (read_buffer_.Length() < num_bytes) {
+      const std::function<void(absl::Status)> read_callback =
+          [this, num_bytes, &read_callback](absl::Status status) {
+            if (!status.ok()) {
+              pending_read_buffer_.Clear();
+              temporary_read_buffer_.Clear();
+
+              grpc_core::MutexLock lock(&read_mutex_);
+              read_result_ = status;
+              read_waker_.Wakeup();
+            } else {
+              while (temporary_read_buffer_.Count() > 0u) {
+                pending_read_buffer_.Append(temporary_read_buffer_.TakeFirst());
+              }
+
+              if (read_buffer_.Count() + pending_read_buffer_.Count() <
+                  num_bytes) {
+                /// A further read is needed.
+                endpoint_->Read(read_callback, &temporary_read_buffer_,
+                                nullptr /* uses default arguments */);
+              } else {
+                while (pending_read_buffer_.Count() > 0u) {
+                  grpc_slice_buffer_add(
+                      read_buffer_.c_slice_buffer(),
+                      pending_read_buffer_.TakeFirst().c_slice());
+                }
+
+                grpc_core::MutexLock lock(&read_mutex_);
+                read_result_ = status;
+                read_waker_.Wakeup();
+              }
+            }
+          };
+      temporary_read_buffer_.Clear();  /// may be redundant
+      endpoint_->Read(read_callback, &temporary_read_buffer_,
+                      nullptr /* uses default arguments */);
+    } else {
+      read_result_ = absl::OkStatus();
+    }
+
+    return [this, num_bytes]()
+               -> grpc_core::Poll<absl::StatusOr<grpc_core::SliceBuffer>> {
+      grpc_core::MutexLock lock(&read_mutex_);
+      if (!read_result_.has_value()) {
+        read_waker_ = grpc_core::Activity::current()->MakeNonOwningWaker();
+        return grpc_core::Pending();
+      } else if (!read_result_->ok()) {
+        const absl::Status ret = *read_result_;
+        read_result_.reset();
+
+        return ret;
+      } else {
+        grpc_core::SliceBuffer ret;
+        grpc_slice_buffer_move_first(read_buffer_.c_slice_buffer(), num_bytes,
+                                     ret.c_slice_buffer());
+
+        read_result_.reset();
+        return ret;
+      }
+    };
+  }
+
+  auto ReadSlice(size_t num_bytes) {
+    grpc_core::MutexLock lock(&read_mutex_);
+
+    /// Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+
+    if (read_buffer_.Length() < num_bytes) {
+      /// TODO: handle potential integer overflow
+      /// TODO: evaluate the lifespan of `read_args`
+      const struct grpc_event_engine::experimental::EventEngine::Endpoint::
+          ReadArgs read_args = {static_cast<int64_t>(num_bytes)};
+
+      const std::function<void(absl::Status)> read_callback =
+          [this, num_bytes, read_args, &read_callback](absl::Status status) {
```

Comment:
const auto (no need for type erasure here)

also... consider moving `num_bytes`, `read_callback` into members on the `PromiseBasedEndpoint`: storing more than one pointer on a callback will force a memory allocation when we call `Endpoint::Read` and convert this callback into an `AnyInvocable`.